Lib/importlib/_bootstrap.py

```diff
@@ -824,10 +824,16 @@ def _module_repr_from_spec(spec):
     """Return the repr to use for the module."""
     name = '?' if spec.name is None else spec.name
     if spec.origin is None:
-        if spec.loader is None:
+        loader = spec.loader
+        if loader is None:
             return f'<module {name!r}>'
+        elif (
+            _bootstrap_external is not None
+            and isinstance(loader, _bootstrap_external.NamespaceLoader)
+        ):
+            return f'<module {name!r} (namespace) from {list(loader._path)}>'
         else:
-            return f'<module {name!r} (namespace) from {list(spec.loader._path)}>'
+            return f'<module {name!r} ({loader!r})>'
```

Comment:
Alternately, to avoid the `_boostrap_external` check:

```py
        else:
            path = getattr(loader, '_path', None)
            if path is not None:
                return f'<module {name!r} (namespace) from {list(path)}>'
            else:
                return f'<module {name!r} ({loader!r})>'
```

or, better, to fall back to the base repr in any unusual case:

```py
        else:
            try:
                path = list(loader._path)
            except Exception:
                return f'<module {name!r} ({loader!r})>'
            else:
                return f'<module {name!r} (namespace) from {path}>'
```

Reply:
Hmm, the consensus on the issue was that an `isinstance()` check would be more reliable than a `hasattr(loader, "_path")` check or a `getattr(loader, "_path", None)` check: https://github.com/python/cpython/issues/112414#issuecomment-1826440461

Reply:
That's fine with me (though it is technically a change in behavior); there's nothing to change then. 😄

It might be worth noting in the NEWS entry that `_path` on a loader is no longer reflected in the repr for anything but namespace packages.  (Sometimes people come to rely on internal implementation details like this, even when it's hard for us to imagine why.)

Reply:
We only started reflecting the `_path` attribute in the repr in 3.12.0, so _hopefully_ we've made this change quickly enough that nobody will have started depending on the `_path` attribute being reflected in the repr ;)

Reply:
Then updating the NEWS entry probably isn't worth doing either, so I'm calling this good.